### ISSUE TITLE
[BUG] quantized_matmul gives non reproducible NaNs

### ISSUE THREAD

**Issue by tyeah** [ml-explore/mlx#896]
**Describe the bug**
I have a model using quantized_matmuls. It works fine in inference. And it works in when I fine-tune the lora weights in that model in the first iteration, but starting from the second iteration, loss becomes NaN. 

I cache intermediate values and set a breakpoint after the optimizer update. I found the output of the first quantuzed_matmul is NaN (the second operation in my model). Now I start trying different stuff in this pdb session:
1. I feed the same inputs (including parameters in the QuantizedLinear module) to mx.quantized_matmul using cpu stream, it works fine. 
2. I dump the inputs to file and load bad, and run it on gpu, it works fine as well. 
3. If I still use the same inputs and run quantized_linear on gpu again, NaNs still.

**To Reproduce**

It's hard to reproduce since the project is a bit complicated.. And as mentioned previously, it cannot be reproduced with standalone mx.quantized_matmul with saved tensors. However, it is reproducible consistently with my training script.

I will appreciate any general suggestions or tips to further debug this issue.

For example, this symptom seems to suggest that there is some state/cache in the gpu pipeline? If so, maybe there is a way to "force reset" any possible state?

Thank you in advance!

**Expected behavior**
Identical or similar result as the cpu quantized_matmul.

**Desktop (please complete the following information):**
 - OS Version: [MacOS 14.1]
 - Version [0.8.0]

**Additional context**
I'm using float16 for inputs and non-quantized parameters.

**Comment by awni** [comment]
Coming up with a simple way to reproduce the issue will be the most useful thing to do to debug this both for yourself and for us to be able to help you.

> I dump the inputs to file and load bad, and run it on gpu, it works fine as well.
> If I still use the same inputs and run quantized_linear on gpu again, NaNs still.

That sounds bad, it should not be the case that any GPU op depends on the state of the machine other than the input values themselves. 

Can you give more input on the op you are running:
- op name
- arguments (bits, group size, etc)

>  it is reproducible consistently with my training script.

Does it show up early? Can you share even a subset of the script that reproduces it?

**Comment by tyeah** [comment]
Hi Awni, thanks for the prompt reply!

I tried to dig deep to carve a reproducible part of the script. I am able to get the following script. I do not know if it reproduces exactly the same issue, but at least:
(1) GPU consistently generate some NaNs, while CPU matches the no-NaN results from dequantized version.
(2) GPU output is all NaNs in some cases, and partially NaNs in other cases.

```python
import mlx.core as mx

mx.random.seed(0)
dtype = mx.float16

weight = mx.random.randint(46077963, 46077963 + 100000, (1, 80), dtype=dtype)
scales = mx.array([[0.00012, 0.00016, 0.00014, 0.00015, 0.00014]], dtype=dtype)
scales = mx.array([[-0.014, -0.021, -0.017, -0.017, -0.018]], dtype=dtype)
x = mx.array([[-1.25195]], dtype=dtype)

y_gpu = mx.quantized_matmul(
        x,
        weight,
        scales=scales,
        biases=biases,
        transpose=False,
        group_size=64,
        bits=8,
        stream=mx.gpu)

y_cpu = mx.quantized_matmul(
        x,
        weight,
        scales=scales,
        biases=biases,
        transpose=False,
        group_size=64,
        bits=8,
        stream=mx.cpu)

print(y_gpu)
# some or all NaNs
print(y_cpu)
# meaningful result

weight_deq = mx.dequantize(weight, scales, biases, 64, 8)
y_deq = x @ weight_deq
print(y_deq)
# meaningful result
print(mx.max(mx.abs(y_cpu - y_deq)))
# ~ 1.5e-5
```
I have not been able to reproduce the "same pdb session, cpu / gpu with loaded variables work, gpu with original variables fails" scenario. But this script does suggest some issue within the quantized_matmul op? I'm hoping they share the same underlying root cause.

Also, while dumping these scale/bias numbers, I do notice the scales are fairly small for float16 (although they come from real models). Could that be an issue? My 2 cents are I can multiply the scales with a big number to put it into a reasonable range (which requires some tuning), or simply use float32 (which could lead to more computation). But I don't know if these are good solutions to this problem, or should I try to do something ad-hoc like this.

**Comment by awni** [comment]
The code you sent doesn't run for me... `quantized_matmul` needs `uint32` `weight`s.  Could you double check it please?

**Comment by tyeah** [comment]
Sorry for the copy paste. Try the updated version below:

```python
import mlx.core as mx

mx.random.seed(0)
dtype = mx.float16

weight = mx.random.randint(46077963, 46077963 + 100000, (1, 80), dtype=mx.uint32)
scales = mx.array([[0.00012, 0.00016, 0.00014, 0.00015, 0.00014]], dtype=dtype)
biases = mx.array([[-0.014, -0.021, -0.017, -0.017, -0.018]], dtype=dtype)
x = mx.array([[-1.25195]], dtype=dtype)

y_gpu = mx.quantized_matmul(
        x,
        weight,
        scales=scales,
        biases=biases,
        transpose=False,
        group_size=64,
        bits=8,
        stream=mx.gpu)

y_cpu = mx.quantized_matmul(
        x,
        weight,
        scales=scales,
        biases=biases,
        transpose=False,
        group_size=64,
        bits=8,
        stream=mx.cpu)

print(y_gpu)
# some or all NaNs
print(y_cpu)
# meaningful result

weight_deq = mx.dequantize(weight, scales, biases, 64, 8)
y_deq = x @ weight_deq
print(y_deq)
print(mx.max(mx.abs(y_cpu - y_deq)))
# ~ 1.5e-5
```

**Comment by awni** [comment]
Ok it runs now and I see the same issue.

**Comment by tyeah** [comment]
Thanks! Let me know if there's anything I can do to help debug.
